### PR TITLE
Improve Copy/Paste

### PR DESCRIPTION
**Pull request by speedplane**
# The Problem

Copy/paste for many PDFs is broken.  Even in [the TraceMonkey example](http://mozilla.github.io/pdf.js/web/viewer.html), try copying and pasting the area highlighted in the image below, and you will get the following text:

```
 informationis available
```

image
Notice that when you paste, there is no space between `information` and `is`.  That is, there is no space between new lines. 

But there are examples that are much worse than the above. In [the PDF at the link here](https://github.com/speedplane/pdf.js/raw/text-extract-with-spaces/test/pdfs/US6205527_page1.pdf), the PDF structure itself has no spaces. That is, whatever software created this PDF, only created PDF objects for the text itself, and did not consider spacing.  Thus, when one copy/pastes anything from this document using pdf.js, they get no spaces in their pasted text. 

It's not just copy/paste that's a problem, but CTRL+F find also does not work.  If you CTRL+F for `information is` in the TraceMonkey example, you will not get any results.
# Other Viewers

Other PDF software gets this right, they figure out that there is a space between words or lines and automatically insert the proper space when the user copies text. Adobe Reader, for example, will not only detect a space and insert it, but it will also detect when a line ends with a hyphen and automatically remove it when doing copy/paste. For example, if you copied the word `comp-ile` in the image above, Reader would actually copy `compile`.  That is probably overkill, but it is interesting that others gave this signifigant amount of thought.

Other software that also gets this right includes [pdf miner](https://github.com/euske/pdfminer), a project that goes to great lengths to properly extract text from PDFs. The Chrome viewer (Foxit) also gets this right (it also does hyphens right too).
# The Solution

This commit aims to resolve these issues by automatically detecting when we need spaces, and inserting them into the text chunks where appropriate.  The insertion occurs within the worker, so not only will copy/paste work, but `getTextContent()` will now also return text with proper spacing. Accordingly, any headless text extraction will also benefit from this PR.

Unfortunately, for PDFs that do not have the `wordSpacing` text state set, this feature must resort to using a bit of a heuristic, but I do not see anyway around that. 

I also added a test which hooks into the `getTextContent` function.

### CODE REVIEW

**Comment by existentialism** [comment]
Nice work! Early run through looks great, will testing more thoroughly and will report anything odd. I had started something similar a while back to try and improve copy/paste, glad to see I was on the right path :+1:

**Comment by speedplane** [comment]
@timvandermeij Hi Tim - can you make a build of this when you get a chance?

**Comment by timvandermeij** [comment]
Sure, no problem!

/botio-linux preview

**Comment by timvandermeij** [comment]
Hm, I see no change at all for the Tracemonkey paper. I tried copying the part in your image, but it still copies without a space at the newline for me.

**Comment by speedplane** [comment]
@timvandermeij I see that too with the built viewer, but it works in my local checkout. In my local checkout, I have PR #5545, which has a change relating to the text layer's line-height. I bet that's what's fixing this case. I'll look into the best fix here (either incorporating the small bit from #5545 or changing this code slightly). I'll also add a test for the tracemonkey PDF specifically.

**Comment by speedplane** [comment]
@timvandermeij Hi Tim... I just tried it with the viewer you built, and it works fine.

**Comment by speedplane** [comment]
@timvandermeij I just updated the commit, adding a test specifically for the tracemonkey paper. It's passing for me.

**Comment by timvandermeij** [comment]
Still no luck for me. If I copy the text "information is available" from the first paragraph, I get:

<pre>
information
is available
</pre>

with the current master and

<pre>
information
is available
</pre>

with the build above, i.e. both are the same. Also if I copy both texts and paste them in for example the Firefox URL bar, both appear as "informationis available", i.e. without a space before "is". You cannot reproduce that?

**Comment by speedplane** [comment]
@timvandermeij Very strange. First off, where are you testing master? I assumed that [the demo page](http://mozilla.github.io/pdf.js/web/viewer.html) was master, but is that incorrect?

I don't get a line break in either of them.

Here is my result copied from http://107.21.233.14:8877/7b23cd1eedc95f6/web/viewer.html
```
information is available
```

And here is the result copied from http://mozilla.github.io/pdf.js/web/viewer.html
```
informationis available
```

I'm using Chrome on Windows. Could it be a browser thing?

**Comment by speedplane** [comment]
Also, if you have time, could you take a look at the DOM in both and see what you get?

This is the DOM from http://107.21.233.14:8877/7b23cd1eedc95f6/web/viewer.html:
image

And here is the DOM from http://mozilla.github.io/pdf.js/web/viewer.html:
image

Notice there is an extra space in the first.

**Comment by timvandermeij** [comment]
I do get the space in the DOM, but for some reason it doesn't copy it for both links you posted. I'm really confused about that: is it a browser bug perhaps? I'm using Firefox 36.0.1 on Windows 7 x64.

**Comment by timvandermeij** [comment]
It does appear to be something with the browser. Take http://www.sersc.org/journals/IJSIP/vol4_no3/5.pdf for example. At the bottom of the first page it says "established by ISO". If you select the text carefully, you can also select a space character behind ISO. Copy/pasting that does give "ISO " with a space, but if you also take the next line the space disappears when you copy/paste. Really weird...

**Comment by speedplane** [comment]
Maybe it isn't a browser bug. Maybe Firefox is actually smarter than the others.  As you pointed out, in the TraceMonkey paper, Firefox is inserting a line-break between lines. There are no line-break characters within the DOM itself, so Firefox must be doing something to detect it and insert it.  Other browsers do not insert line-breaks. 

I would bet that as part of that logic, when it detects a line break in selected text, it strips the end of the line of any whitespace.

**Comment by timvandermeij** [comment]
Good point! That would mean that this patch would not have much effect in Firefox, but it will help for other browsers.

**Comment by speedplane** [comment]
Try copying and pasting text in the following jsfiddle: https://jsfiddle.net/4wwd654y/

In Chrome, if you try copying and pasting, the entire thing, you get: 
```
Line OneLine Two
```
I put money on the fact that in Firefox you would get:
```
Line One
Line Two
```

**Comment by speedplane** [comment]
Regarding your second point: this patch would not have an effect on the Firefox Tracemonkey paper. However, it would still fix [messed up PDFs like this one in firefox](https://github.com/speedplane/pdf.js/raw/text-extract-with-spaces/test/pdfs/US6205527_page1.pdf).

**Comment by timvandermeij** [comment]
You're right, that is what is happening in Firefox. However, for the PDF you linked in the comment just above this one, I also see no difference in Firefox. Both https://mozilla.github.io/pdf.js/web/viewer.html and http://107.21.233.14:8877/7b23cd1eedc95f6/web/viewer.html copy any text without spaces for me. The DOM shows the spaces in each `div`, but since they are separate `div`s, Firefox strips those spaces too. I think we'll have to look into this later, but nevertheless this seems like a great patch for other browsers and for when we figure out what exactly is different in Firefox.

File: src/core/evaluator.js [hunk 1]
```diff
@@ -1037,8 +1037,66 @@ var PartialEvaluator = (function PartialEvaluatorClosure() {
...
+        }
+        var fontHeight = Math.sqrt((tx[2] * tx[2]) + (tx[3] * tx[3]));
+        var fontAscent = fontHeight;
+        if (font.ascent) {
+          fontAscent = font.ascent * fontAscent;
+        } else if (font.descent) {
+          fontAscent = (1 + font.descent) * fontAscent;
+        }
+        return {
+          x : (angle === 0) ? tx[4] :(tx[4] + (fontAscent * Math.sin(angle))),
```
Comment (timvandermeij) [inline src/core/evaluator.js]: Nit: change this line to `x: (angle === 0 ? tx[4] : tx[4] + (fontAscent * Math.sin(angle))),` to get the spaces and parentheses right

File: test/unit/text_extract.js [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+  var pagePromise = new Promise(function (resolve) {
+    resolvePromise = resolve;
+  });
+  PDFJS.getDocument(pdfURL).then(function(doc) {
+    doc.getPage(1).then(function(data) {
+      resolvePromise(data);
+    });
+  });
+  var page = {
+    promise : pagePromise,
```
Comment (timvandermeij) [inline test/unit/text_extract.js]: Nit: remove space before `:` here and on the next line

**Comment by speedplane** [comment]
Fixed the nits. 

Also, I did some testing of this PR in Firefox and other browsers.

**TLDR**: This PR fixes a number of issues in Chrome and IE and fixes a CTRL+F issue in Firefox. Also, I spotted a real browser bug in Firefox.

---
PDF.js uses absolutely positioned divs in its text layer. Selecting, copying, and pasting absolutely positioned divs is not an easy problem and browsers work differently. Firefox tries to be smart about copy/paste. In most browsers, if you select, copy, and paste more than one absolutely positioned div, no space will be put between them. In Firefox, if two absolutely positioned divs are selected, and they are separated by more than some set distance, then firefox inserts a new line in between the two divs.

You can see this by viewing [this new jsfiddle](https://jsfiddle.net/4wwd654y/1/). Ideally, a browser would copy/paste something like the following:
```
Line One
Line Two

Line Three A Line Three B
```
Unfortunately, no browser does that. Chrome and IE will put everything on the same line, without any space between:
```
Line OneLine TwoLine Three ALine Three B
```

Firefox works a bit differently. When you do a copy paste, it adds new lines between each div:
```
Line One
Line Two
Line Three A
Line Three B
```

This PR fixes the issue in Chrome and IE by automatically putting a space at the end of a div when the following div is more than a set distance away.  [This jsfiddle simulates that](https://jsfiddle.net/4wwd654y/2/). Thus, in Chrome and IE, the copy paste will now work nicely.  

Unfortunately, this PR has no effect on Firefox. It will continue to put new lines between divs even though there should not be a new line, i.e., it will always put `Line Three A` and `Line Three B` on the same line. This appears to be a browser bug.

**But wait, there's more.**  Copy/paste is not the only issues addressed in this PR. It also fixes another aspect of the text layer: PDF.js's find feature. For example, if you load the [TraceMonkey example](http://mozilla.github.io/pdf.js/web/viewer.html) (in any browser) and do a CTRL+F for `information is`, the first hit should be in the first paragraph.  However, instead, you get nothing.  If you search for `informationis`, then the first hit appears.  

This PR also fixes this CTRL+F issues because we are inserting a space into the text layer. This is **not** a Firefox only fix, this fix is cross-browser.  

Admittedly, the fix in the Tracemonkey example is subtle.  However, if you view [the messed up PDF which has no spaces](https://github.com/speedplane/pdf.js/raw/text-extract-with-spaces/test/pdfs/US6205527_page1.pdf), the improvement is much more stark. Try loading that PDF and searching for any of the words in the first line of the abstract, such as `disclosed is an`.  Across all browsers, you will get no results in master, but you will get the correct result with this PR.

**Comment by timvandermeij** [comment]
@speedplane Thank you for the analysis! I can confirm that Ctrl+F at line breaks now works properly with your patch applied. In fact, it seems to fix a lot of PDFs with a two-column layout for me. Really nice work!

File: src/core/evaluator.js [hunk 3]
```diff
@@ -1037,8 +1037,66 @@ var PartialEvaluator = (function PartialEvaluatorClosure() {
         } else {
           textChunk.height += Math.abs(height * scaleCtmX * scaleLineX);
         }
+        addSpaceIfNecessary(textChunk, font);
         return textChunk;
       }
+      function getChunkPos(chunk, font) {
```
Comment (timvandermeij) [inline src/core/evaluator.js]: Could we rename this to `getChunkPosition` for readability?

File: src/core/evaluator.js [hunk 4]
```diff
@@ -1037,8 +1037,66 @@ var PartialEvaluator = (function PartialEvaluatorClosure() {
...
+        var tx = chunk.transform;
+        var angle = Math.atan2(tx[1], tx[0]);
+        if (font.vertical) {
+          angle += Math.PI / 2;
+        }
+        var fontHeight = Math.sqrt((tx[2] * tx[2]) + (tx[3] * tx[3]));
```
Comment (timvandermeij) [inline src/core/evaluator.js]: Why not put this directly in `fontAscent`? `fontHeight` is not used after directly assigning it to `fontAscent` below.

File: src/core/evaluator.js [hunk 5]
```diff
@@ -1037,8 +1037,66 @@ var PartialEvaluator = (function PartialEvaluatorClosure() {
...
+        // If the last chunk ends with a space it does not need one.
+        var lastChunk = bidiTexts[bidiTexts.length - 1];
+        if (lastChunk.str.length === 0) {
+          return;
+        }
+        var lastChar = lastChunk.str[lastChunk.str.length - 1];
+        if (lastChar === ' ' || lastChar === '-') {
+          return;
+        }
+        var lastPos = getChunkPos(lastChunk, font);
```
Comment (timvandermeij) [inline src/core/evaluator.js]: Rename `getChuckPos` to `getChunkPosition` here and the line below this one too.

File: src/core/evaluator.js [hunk 6]
```diff
@@ -1037,8 +1037,66 @@ var PartialEvaluator = (function PartialEvaluatorClosure() {
...
+        var newPos = getChunkPos(newChunk, font);
+        var yDiff = Math.abs(lastPos.y  - newPos.y);
+        if (yDiff >= lastChunk.height || yDiff >= newChunk.height) {
+          // On different lines, add a space.
+          lastChunk.str += ' ';
+        } else {
+          var wordSpacing = textState.wordSpacing > 0 ?
```
Comment (timvandermeij) [inline src/core/evaluator.js]: Let's use a regular `if` statement here instead of the ternary operator for readability. I think the ternary operator should only be used if the statement can (reasonably) be put on a single line; this statement including the comments is too long for that. Also applies to `addSpace` below.

File: test/unit/text_extract.js [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+  };
+  waitsForPromiseResolved(pagePromise, function(data) {
+    page.page = data;
+  });
+  return page;
+}
+
+describe('text-extract', function() {
+  it('patent', function () {
+    var page = getPageOneOf('../pdfs/US6205527_page1.pdf');
```
Comment (timvandermeij) [inline test/unit/text_extract.js]: What is the license of this PDF file? Can we use it freely?

**Comment by speedplane** [comment]
Implemented all of your suggestions. Regarding the PDF, it is a U.S. patent, which is a government work, and accordingly there can be no copyright claim. 

Also note, that this should also fix #5697.

File: src/core/evaluator.js [hunk 8]
```diff
@@ -1037,8 +1037,72 @@ var PartialEvaluator = (function PartialEvaluatorClosure() {
...
+        var newPosisition = getChunkPosition(newChunk, font);
+        var yDiff = Math.abs(lastPosition.y  - newPosisition.y);
+        if (yDiff >= lastChunk.height || yDiff >= newChunk.height) {
+          // On different lines, add a space.
+          lastChunk.str += ' ';
+        } else {
+          var wordSpacing = textState.wordSpacing; // Try default wordSpacing.
+          if (wordSpacing === 0) {
+            // Hueristic for wordSpacing
+            wordSpacing = newChunk.width / newChunk.str.length * 0.6;
```
Comment (Snuffleupagus) [inline src/core/evaluator.js]: Since `0.6` is a heuristic value, it should be replaced with a suitably named constant, placed amongst the other ones at the top of [`getTextContent`](https://github.com/mozilla/pdf.js/blob/master/src/core/evaluator.js#L896-L897).

File: src/core/evaluator.js [hunk 9]
```diff
@@ -1037,8 +1037,72 @@ var PartialEvaluator = (function PartialEvaluatorClosure() {
...
+          }
+          var addSpace;
+          if (newPosisition.x >= lastPosition.x) {
+            // Left to right. Add a space if next is past wordSpacing.
+            addSpace = newPosisition.x >= lastPosition.x + lastChunk.width +
```
Comment (Snuffleupagus) [inline src/core/evaluator.js]: Nit: please consider adding parentheses here (and below) to improve the readability of the code.

**Comment by Snuffleupagus** [comment]
Given that the textLayer is currently one of the main performance pain points for a lot of PDF documents, what kind of performance impact does this PR have?
(To help with benchmarking, please see https://github.com/mozilla/pdf.js/wiki/Benchmarking-your-changes)

**Comment by speedplane** [comment]
@Snuffleupagus I tried running the benchmark script, with no success, any idea why?

```
C:\...\pdf.js_speedplane\test>node test.js --browserManifestFile=resources/browser_manifests/browser_manifest.json --statsFile=stats/results/baseline.json --statsDelay=5000 --manifestFile=my_pdfs.json --port=7489
Server running at http://127.0.0.1:7489/

events.js:72
        throw er; // Unhandled 'error' event
              ^
Error: spawn ENOENT
    at errnoException (child_process.js:1001:11)
    at Process.ChildProcess._handle.onexit (child_process.js:792:34)
```

File: src/core/evaluator.js [hunk 10]
```diff
@@ -1037,8 +1037,72 @@ var PartialEvaluator = (function PartialEvaluatorClosure() {
...
+        // Start by calculating the height
+        var fontAscent = Math.sqrt((tx[2] * tx[2]) + (tx[3] * tx[3]));
+        // Then modify to add the ascent
+        if (font.ascent) {
+          fontAscent = font.ascent * fontAscent;
+        } else if (font.descent) {
+          fontAscent = (1 + font.descent) * fontAscent;
+        }
+        return {
+          x : (angle === 0 ? tx[4] : tx[4] + (fontAscent * Math.sin(angle))),
```
Comment (timvandermeij) [inline src/core/evaluator.js]: Nit: change `x :` to `x:` (i.e., remove the space just like below)

File: src/core/evaluator.js [hunk 11]
```diff
@@ -1037,8 +1037,72 @@ var PartialEvaluator = (function PartialEvaluatorClosure() {
...
+        var lastChunk = bidiTexts[bidiTexts.length - 1];
+        if (lastChunk.str.length === 0) {
+          return;
+        }
+        var lastChar = lastChunk.str[lastChunk.str.length - 1];
+        if (lastChar === ' ' || lastChar === '-') {
+          return;
+        }
+        var lastPosition = getChunkPosition(lastChunk, font);
+        var newPosisition = getChunkPosition(newChunk, font);
```
Comment (timvandermeij) [inline src/core/evaluator.js]: Typo: `newPosisition` -> `newPosition`

File: src/core/evaluator.js [hunk 12]
```diff
@@ -1037,8 +1037,72 @@ var PartialEvaluator = (function PartialEvaluatorClosure() {
...
+        var yDiff = Math.abs(lastPosition.y  - newPosisition.y);
+        if (yDiff >= lastChunk.height || yDiff >= newChunk.height) {
+          // On different lines, add a space.
+          lastChunk.str += ' ';
+        } else {
+          var wordSpacing = textState.wordSpacing; // Try default wordSpacing.
+          if (wordSpacing === 0) {
+            // Hueristic for wordSpacing
```
Comment (timvandermeij) [inline src/core/evaluator.js]: Typo: `Hueristic` -> `Heuristic`

File: src/core/evaluator.js [hunk 13]
```diff
@@ -1037,8 +1037,72 @@ var PartialEvaluator = (function PartialEvaluatorClosure() {
...
+            wordSpacing = newChunk.width / newChunk.str.length * 0.6;
+          }
+          var addSpace;
+          if (newPosisition.x >= lastPosition.x) {
+            // Left to right. Add a space if next is past wordSpacing.
+            addSpace = newPosisition.x >= lastPosition.x + lastChunk.width +
+              wordSpacing;
+          } else {
+            // Right to left. Add space if next is before sart.
```
Comment (timvandermeij) [inline src/core/evaluator.js]: Typ: `sart` -> `start` ?

File: test/unit/unit_test.html [hunk 14]
```diff
@@ -51,6 +51,7 @@
   <script src="stream_spec.js"></script>
   <script src="parser_spec.js"></script>
   <script src="api_spec.js"></script>
+  <script src="text_extract.js"></script>
```
Comment (timvandermeij) [inline test/unit/unit_test.html]: Perhaps rename this to `text_layer_spec.js` so we can include all text layer related unit tests in there? It's also more consistent with the names of the other test suites.

**Comment by timvandermeij** [comment]
@speedplane Probably it is trying to spawn a process/application that it cannot find. Run `npm update` to get all dependencies. Probably it is https://github.com/mozilla/pdf.js/blob/master/test/webbrowser.js#L25 that is failing.

**Comment by jazzy-em** [comment]
@speedplane, great work! IMHO, you resolve very important problem!

**Comment by jazzy-em** [comment]
I found some bugs:
1. Small difference in highlighting. There is small shift to the left.
For example, search for 'up' in the TraceMonkey example document
Original:
original
Your:
your

**Comment by jazzy-em** [comment]
2. In this document (http://www.selab.isti.cnr.it/ws-mate/example.pdf) try to search "an example paper".
There will no matches, because there is no space between 'example' and 'paper'.
ex1
--
ex2

**Comment by timvandermeij** [comment]
@speedplane This also needs a rebase onto the current master.

**Comment by danez** [comment]
What is the status of this? Any news?

**Comment by speedplane** [comment]
@danez been very busy with other projects. I want to clean up this PR (and
others that I submitted), but won't realistically get to them until late
August.

On Tue, Jun 30, 2015 at 11:18 AM, Daniel Tschinder <notifications@github.com

**Comment by danez** [comment]
In the case of this [pdf][1] there are chunks between each word that only contain the space itself. 

The text-layer/viewer is not rendering the text-divs as the width of a space is calculated as 0. I first tried to change something there and render these divs with a single space, but then all the scale-calculation cannot be done as of width being zero. 

So I ended up adding spaces to the surrounding words if there are space-chunks in-between to fix that. What do you think about that?

```patch
-        // If the last chunk ends with a space it does not need one.
+
         var lastChunk = bidiTexts[bidiTexts.length - 1];
+
+        // If the last chunk equals a single space concat the space to chunk before the last chunk
+        if (lastChunk.str === ' ' && bidiTexts.length - 2 >= 0) {
+            var beforeLastChunk = bidiTexts[bidiTexts.length - 2];
+            if (beforeLastChunk.str.length === 0) {
+                return;
+            }
+            var lastChar = beforeLastChunk.str[beforeLastChunk.str.length - 1];
+            if (lastChar !== ' ') {
+                beforeLastChunk.str += ' ';
+            }
+            return;
+        }
+
         if (lastChunk.str.length === 0) {
           return;
         }
-        var lastChar = lastChunk.str[lastChunk.str.length - 1];
+
+        // If the last chunk ends with a space it does not need one.
+        lastChar = lastChunk.str[lastChunk.str.length - 1];
         if (lastChar === ' ' || lastChar === '-') {
           return;
         }
```

[1]: https://www.researchgate.net/publication/257019040_Shortcomings_of_classical_phenological_forcing_models_and_a_way_to_overcome_them

**Comment by danez** [comment]
I also noticed that the selection is now off in the web-viewer, because the space is added to the div, but the scale is wrong.
You see the selection ends inside the character. (can also be seen in the image from jazzy-em above)
bildschirmfoto 2015-07-01 um 12 09 52

This of course gets worse with my patch from above.

File: src/core/evaluator.js [hunk 15]
```diff
@@ -1037,8 +1040,73 @@ var PartialEvaluator = (function PartialEvaluatorClosure() {
...
+        // If the new chunk starts with a space, it does not need one.
+        if (newChunk.str[0] === ' ' || newChunk.str[0] === '-') {
+          return;
+        }
+        if (bidiTexts.length === 0) {
+          return;
+        }
+        // If the last chunk ends with a space it does not need one.
+        var lastChunk = bidiTexts[bidiTexts.length - 1];
+        if (lastChunk.str.length === 0) {
```
Comment (Hengjie) [inline src/core/evaluator.js]: Change to `lastChunk && lastChunk.str.length === 0` since there could be no `lastChunk`. This avoids it throwing an error.

**Comment by Vad1mo** [comment]
hello,
what is the status of the pull request. As I see it even if it doesn't solve the problem as desired it is already a big step forward! Can we not merge it into master?

**Comment by timvandermeij** [comment]
It needs to be rebased onto the current master and reviewed before we can land this.

**Comment by fbender** [comment]
@speedplane any idea when you will get to this?

**Comment by adarowiec** [comment]
almost 3 years and it's still open? excuse me if I'm missing something, but it seems that this issue is still present, at least in my domain - and I was wondering if it's something specific or global, that's all I need

**Comment by timvandermeij** [comment]
Closing since this was never completely finished and it's not in a mergeable state. However, the associated issue will remain open with a reference to this pull request so the work can be continued later on.
